src/unix/async.c

```diff
@@ -38,6 +38,33 @@
 #include <sys/eventfd.h>
 #endif
 
+#if UV__KQUEUE_EVFILT_USER
+static uv_once_t kqueue_runtime_detection_guard = UV_ONCE_INIT;
+static int kqueue_evfilt_user_support = 1;
+
+
+static void uv__kqueue_runtime_detection(void) {
+  int kq;
+  struct kevent ev[2];
+  struct timespec timeout = {1, 0};
+
+  /* Perform the runtime detection to ensure that kqueue with
+   * EVFILT_USER actually works. */
+  kq = kqueue();
+  EV_SET(ev, UV__KQUEUE_EVFILT_USER_IDENT, EVFILT_USER,
+         EV_ADD | EV_CLEAR, 0, 0, 0);
+  EV_SET(ev + 1, UV__KQUEUE_EVFILT_USER_IDENT, EVFILT_USER,
+         0, NOTE_TRIGGER, 0, 0);
+  if (kevent(kq, ev, 2, ev, 1, &timeout) < 1 || 
+      ev[0].filter != EVFILT_USER ||
+      ev[0].ident != UV__KQUEUE_EVFILT_USER_IDENT)
```

Comment:
Is all this necessary, or would it suffice just to check if the filter is valid (or if it causes EINVAL)
```suggestion
  EV_SET(ev, 0, EVFILT_USER, EV_ADD | EV_CLEAR, 0, 0, 0);
  if (kevent(kq, ev, 1, NULL, 0, NULL) != 0)
```

Reply:
I am still a bit confused by the complexity here, though it does seem correct now. The things you are checking for `ev[0].filter` is guaranteed to be `EVFILT_USER` here and `ev[0].ident` is guaranteed to be `UV__KQUEUE_EVFILT_USER_IDENT` on every version of kqueue (either with `EV_ERROR` set on an unsupported kernel or not set if it worked). Anything else appearing here would be a pretty significant kernel bug (or userspace stack corruption).